content/ytt/docs/latest/data-values-schema-migration-guide.md

```diff
@@ -0,0 +1,201 @@
+#@ load("@ytt:library", "library")
+#@ load("@ytt:template", "template")
+
+#@ lib = library.get("lib").with_data_values(data.values)
+--- #@ template.replace(lib.eval())
+```
+Using the same `values-1.yml` and `values-2.yml` files from the [multiple data values files schema migration example above](#multiple-data-values-files).
+Migrating to schema happens one library at a time. Let's start with the root library, which includes everything at and below the file system level where the `ytt` invocation was called, not including the `_ytt_lib` folder:
+```bash
+.
+└── config
+    ├── config.yml
+    ├── values-1.yml
+    └── values-2.yml
+```
+As seen in the [previous example](#multiple-data-values-files), `values-1.yml` and `values-2.yml` can be combined and replaced with a single schema file:
+`values-schema.yml`:
+```yaml
+#@data/values-schema
+---
+key1: myVal
+key2: 8088
+key3:
+  host: registry.dev.io
+  port: 8080
+```
+Now we have migrated the root library to use schemas, and the `ytt` invocation will succeed as the same as before, but the private library, "lib" is not yet migrated to schema. 
+```bash
+$ tree .
+.
+└── config
+    ├── config.yml 
+    ├── values-schema.yml
+    └── _ytt_lib
+        └── lib
+            ├── service.yml
+            └── values.yml
+```
+Migrating a private library to use schemas involves the same process as the root library. You can narrow the context to just the children of the `_ytt_lib` directory:
+```bash
+└── _ytt_lib
+    └── lib
+        ├── service.yml
+        └── values.yml
+```
+Now simply follow the steps in either of the previous examples to migrate the private library to use schemas.
+
+## How do I provide default values for an array?
+Arrays in schemas have [special behavior](lang-ref-ytt-schema.md#inferring-defaults-for-arrays), their default value is empty, but their inferred type is from the value in the schema document. 
+Arrays must have one and only one item, which provides the inferred type of that data value. The example below shows how to define an array in a schema and them provide default values via a data values file.
+
+`values-schema.yml`:
+```yaml
+#@data/values-schema
+---
+key: 
+- host: ""
+  port: 0
+  transport: ""
+  insecure_disable_tls_validation: false
+```
+
+`values.yml`:
+```yaml
+#@data/values
+---
+key:
+- host: registry.dev.io
+  port: 8080
+  transport: tcp
+```
+Providing these files along with a template file containing `key: #@ data.values.key` will produce:
+```yaml
+key:
+- host: registry.dev.io
+  port: 8080
+  transport: tcp
+  insecure_disable_tls_validation: false
+```
+
+
+## How do I mark a section of Data Values as "optional"?
+The [@schema/nullalble](lang-ref-ytt-schema.md#schemanullable) annotation can be used to override the defaults of the node it annotates.
+
+Use this annotation on a node if `null` is valid input, or you want the default value for that data value to be `null`. (This annotation cannot be use on arrays.)
```

Comment:
A combination of suggestions:

- lead with describing the user's experience.
- in order to communicate how the concept of "optional" is realized, I think we need to show _both_ how the section is annotated _and_ how it is used.
- seems like an example would be the most terse/effective way to show that.
- that array elements can't be annotated as null is better described in reference material.

```suggestion
Sometimes your configuration includes a section of Data Values that are not typically used or in some way optional.

The [@schema/nullable](lang-ref-ytt-schema.md#schemanullable) annotation can be used to mark such a section as `null`.

        ```yaml
        #@schema/nullable
        aws:
          username: admin
          password: "1234"
        ```

The default value for `data.values.aws` will be `null`.

`null` is falsey.  Templates using that optional data value can perform a simple check for whether the section was specified:

        ```yaml
        #@ load("@ytt:data", "data")
        ---
        #@ if/end data.values.aws:
        s3_creds:
          user: #@ data.values.aws.username
          password: #@ data.values.aws.password
        ```
```

Note: the code examples do _not_ have to be indented... I had to do that in order [to get the full "suggestion" to display in this here GitHub comment](https://github.com/vmg/redcarpet/issues/208#issuecomment-14738077).